pkg/apis/admissionregistration/types.go

```diff
@@ -115,6 +115,234 @@ const (
+	// PolicyName references a ValidatingAdmissionPolicy name which the policyBinding binds to.
+	// if the resource referred to by PolicyName does not exist, this binding is considered mis-configured and will be ignored
+	// Required.
+	PolicyName string
+
+	// ParamName specifies the parameter resource used to configure the admission control policy.
+	// It should point to a Custom Resource which is created out of the CRD specified in ParamSource of ValidatingAdmissionPolicy it binded.
+	// If the resource referred to by ParamName does not exist, this binding is considered mis-configured and the FailurePolicy of the policy definition applied.
+	// +optional
+	ParamName string
+
+	// MatchResources declares what resources match this binding and will be validated by it.
+	// +optional
+	MatchResources *MatchResources
+}
+
+// MatchResources decides whether to run the admission control policy on an object based
+// on whether it meets the match criteria.
+// The exclude rules take precedence over include rules (if a resource matches both, it is excluded)
+type MatchResources struct {
+	// NamespaceSelector decides whether to run the admission control policy on an object based
+	// on whether the namespace for that object matches the selector. If the
+	// object itself is a namespace, the matching is performed on
+	// object.metadata.labels. If the object is another cluster scoped resource,
+	// it never skips the admission control policy.
+	//
+	// See
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels
+	// for examples of label selectors.
+	//
+	// Default to the empty LabelSelector, which matches everything.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector
+	// ObjectSelector decides whether to run the validation based on if the
+	// object has matching labels. objectSelector is evaluated against both
+	// the oldObject and newObject that would be sent to the cel validation, and
+	// is considered to match if either object matches the selector. A null
+	// object (oldObject in the case of create, or newObject in the case of
+	// delete) or an object that cannot have labels (like a
+	// DeploymentRollback or a PodProxyOptions object) is not considered to
+	// match.
+	// Use the object selector only if the webhook is opt-in, because end
+	// users may skip the admission webhook by setting the labels.
+	// Default to the empty LabelSelector, which matches everything.
+	// +optional
+	ObjectSelector *metav1.LabelSelector
+	// ResourceRules describes what operations on what resources/subresources the ValidatingAdmissionPolicy matches.
+	// +optional
+	ResourceRules []RuleWithOperations
+	// ExcludeResourceRules describes what operations on what resources/subresources the ValidatingAdmissionPolicy should not care about.
+	// +optional
+	ExcludeResourceRules []RuleWithOperations
+	// ObjectName specifies the object name which the admission control policy should validate on.
+	// +optional
+	ObjectName []string
```

Comment:
the generateName is still in the REST code, that happens after the mutation and admission per [@sttts docs
](https://www.slideshare.net/sttts/kubernetes-api-codebase-tour)
<img width="847" alt="image" src="https://user-images.githubusercontent.com/6450081/199345396-7ce38fcf-af4d-4296-be73-c0017cecfe19.png">


I also double checked it with the following test, and it shows that Name is empty in admission
```
diff --git a/test/integration/apiserver/admissionwebhook/load_balance_test.go b/test/integration/apiserver/admissionwebhook/load_balance_test.go
index 9423b0f0dc1..77bf711951c 100644
--- a/test/integration/apiserver/admissionwebhook/load_balance_test.go
+++ b/test/integration/apiserver/admissionwebhook/load_balance_test.go
@@ -265,6 +265,7 @@ func newLoadBalanceWebhookHandler(recorder *connectionRecorder) http.Handler {
                if err := json.Unmarshal(review.Request.Object.Raw, pod); err != nil {
                        http.Error(w, err.Error(), 400)
                }
+               fmt.Println("DEBUG", pod.Name, pod.GenerateName)
 
                // When resetting between tests, a marker object is patched until this webhook
                // observes it, at which point it is considered ready.
```

Reply:
Thank you for confirming!^^

Reply:
And thanks for sharing this the awesome pic :)

Reply:
> the generateName is still in the REST code, that happens after the mutation and admission per [@sttts docs](https://www.slideshare.net/sttts/kubernetes-api-codebase-tour)

Just after mutating admission, right? validating admission should still be able to observe the actual (maybe generated) name of the object being persisted.

Coming after mutating admission is unfortunate since it means the semantics of this field will be different between validating policies and any future mutating admission policy we come up, and that a mutating policy guarding objects named `my-special-object` might not intercept an object with `generateName: "my-special-object"` which could still produce an object named `my-special-object` depending on the backing strategy's GenerateName impl.